distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,856 @@
+import asyncio
+import collections
+import logging
+import socket
+import struct
+import weakref
+
+try:
+    import ssl
+except ImportError:
+    ssl = None  # type: ignore
+
+import dask
+
+from ..utils import ensure_ip, get_ip, get_ipv6, nbytes
+from .addressing import parse_host_port, unparse_host_port
+from .core import Comm, CommClosedError, Connector, Listener
+from .registry import Backend
+from .utils import ensure_concrete_host, from_frames, to_frames
+
+logger = logging.getLogger(__name__)
+
+
+_COMM_CLOSED = object()
+
+
+def coalesce_buffers(buffers, target_buffer_size=64 * 1024, small_buffer_size=2048):
+    """Given a list of buffers, coalesce them into a new list of buffers that
+    minimizes both copying and tiny writes.
+
+    Parameters
+    ----------
+    buffers : list of bytes_like
+    target_buffer_size : int, optional
+        The target intermediate buffer size from concatenating small buffers
+        together. Coalesced buffers will be no larger than approximately this size.
+    small_buffer_size : int, optional
+        Buffers <= this size are considered "small" and may be copied. Buffers
+        larger than this may also be copied if the total message length is less
+        than ``target_buffer_size``.
+    """
+    # Nothing to do
+    if len(buffers) == 1:
+        return buffers
+
+    # If the whole message can be sent in <= target_buffer_size, always concatenate
+    if sum(map(len, buffers)) <= target_buffer_size:
```

Comment:
Would suggest using `nbytes` (`from distributed.utils import nbytes`) instead of `len`. If `buffers` contain objects that have a `shape` or a `dtype` that is larger than a byte, `len` will give a smaller value than expected

Reply:
Thanks, nice catch. The original implementation did this correctly, but when copying this up to a new function I missed this invariant. Fixed.